### ISSUE TITLE
Trusted Publishing: Support self-hosted GitLab instances

### ISSUE THREAD

**Issue by di** [pypa/warehouse#15838]
From https://github.com/pypi/warehouse/issues/13575:

> Since the `iss` will be whatever the URL of the GitLab instance is, I'm assuming that for this provider we're not supporting self-hosted instances, right? That is, for now only `"iss": "https://gitlab.com"` will be supported

Scanning the rest of this Issue quickly I didn't see a direct reply on this (apologies if I missed it). I don't have any technical experience on this issue, so is self-hosted GitLab instances something that would be feasible to support in the future? I'm specifically interested in CERN's GitLab instance (c.f. https://github.com/di/id/issues/216) as there are multiple projects there that publish to PyPI where we'd like to transition to using Trusted Publishers.

_Originally posted by @matthewfeickert in https://github.com/pypi/warehouse/issues/13575#issuecomment-2072342830_

**Comment by di** [comment]
I think it would be technically possible to support a self-hosted instance like this (I see that https://gitlab.cern.ch/.well-known/openid-configuration and https://gitlab.cern.ch/oauth/discovery/keys are both publicly available, which is all we need for verification), the real question is what the process by which we would add support for all these one-off issuers.

I think one thing we could do here would be to allow the user to optionally configure the `iss` field as well. My main concern would be that this would allow anyone to essentially masquerade as a GitLab instance, and publish from anywhere, which would give me a little less confidence in the security of the publish event as a user.

Another option is that we allow-list certain issuers for projects in certain organizations, and manually handle these on a case-by-case basis.

Open to ideas though!

**Comment by woodruffw** [comment]
> I think one thing we could do here would be to allow the user to optionally configure the `iss` field as well. My main concern would be that this would allow anyone to essentially masquerade as a GitLab instance, and publish from anywhere, which would give me a little less confidence in the security of the publish event as a user.

Yeah, I think this poses a decent risk 😅. I think a variant of this came up with self-hosted GitHub Enterprise users as well, and IIRC my thoughts there were:

1. Trusted publishing offers diminishing security returns for smaller IdPs: a big part of the security model/value for trusted publishing is the idea that each trusted IdP is large, has on-call staff, has key revocation/rotation policies that they can apply in the event of an emergency, etc. These _might_ be true for smaller IdPs as well, but they're less guaranteed. In the setting of smaller IdPs, IMO manually configured API tokens actually provide better security properties (since revocation of publishing credentials doesn't potentially require handling a rarely-interacted-with internal IdP service).
2. Once the issuer becomes configurable, the "shape" of each OIDC JWT becomes malleable: PyPI would need to be able to distinguish GitHub-looking self-hosted IdPs from GitLab-looking ones, etc., since the claim sets aren't consistent. This also poses a timeseries challenge, e.g. GitHub deploys a claim change to their central IdP to fix a security problem, but we can't respect it until all self-hosted GitHub instances roll their IdP service over as well.

With those being said, I think the allow-list on an organization basis _could_ work! But I think it would require some architectural changes to the current implementation, particularly around claim flexibility 🙂

**Comment by matthewfeickert** [comment]
> Another option is that we allow-list certain issuers for projects in certain organizations, and manually handle these on a case-by-case basis.

This seems the most reasonable, as otherwise there could easily be abuse. I am mindful however that this now requires the maintainer team to now be responsible for a growing list of issuers. Would it help having self-hosted instances that want to use Trusted Publishers self identify for vetting through making a PR to the system that controls the allow-list? Or does this not actually decrease the maintenance burden much?

**Comment by woodruffw** [comment]
> Would it help having self-hosted instances that want to use Trusted Publishers self identify for vetting through making a PR to the system that controls the allow-list? Or does this not actually decrease the maintenance burden much?

I think this would help a bit, but the bulk of the maintenance burden will (unfortunately) probably still be papering over the small differences between each IdP.

I suppose we could reduce the burden of that by enforcing a baseline set of claims for each "shape" of IdP (e.g. GitHub, GitLab) via `.well-known/openid-configuration` on each issuer, but GitLab at minimum has an outstanding issue with that configuration being incomplete: https://gitlab.com/gitlab-org/gitlab/-/issues/428061

**Comment by nickveenhof** [comment]
@woodruffw I'm at a GitLab hackathon (and GitLab team member here) and we're trying to figure out what exactly you would need in the openid-configuration file to move this forward. From how I understand it, that would not necessarily solve any of the verification issues that are being discussed above. Any pointers how we can move forward here after we solve the issue you list? Most likely the warehouse software also need to be modified to have some sort of approval process for self-managed instances?

**Comment by di** [comment]
> I suppose we could reduce the burden of that by enforcing a baseline set of claims for each "shape" of IdP (e.g. GitHub, GitLab) via `.well-known/openid-configuration` on each issuer, but GitLab at minimum has an outstanding issue with that configuration being incomplete: https://gitlab.com/gitlab-org/gitlab/-/issues/428061

I would expect claims from self-managed instances to be roughly the same as the primary hosted instance (although it would be nice if the well-known claims were accurate so we could verify this in advance).

**Comment by woodruffw** [comment]
Hey @nickveenhof! Sorry for the belated response -- I was away from a computer when you sent your original message, and this got lost in the stack.

You're correct that this wouldn't solve the verification issues themselves, since the JWTs issued by a given IdP are the ultimate source of ground truth about the identity being presented.

However, it _would_ allow us to build a cleaner UX for self-hosted instances: one simple way for us to enable self-hosted support is by having people give us a single well-known configuration URL, which we could then validate for the basic "shape" of expected forthcoming OIDC credentials. Similarly, it would help us diagnose and present better diagnostics when validation requirements change (e.g. if we strengthen the constraints on GitLab-shaped publishers, but some self-hosted instances haven't updated to contain newly required claims yet).

**Comment by nejch** [comment]
Hey @woodruffw @di @nickveenhof!

[python-gitlab](https://github.com/python-gitlab) maintainer here and member of the [code.siemens.com](https://about.gitlab.com/customers/siemens/) team. I started https://gitlab.com/gitlab-org/gitlab/-/merge_requests/170072 which I think should remove the blocker upstream issue you linked above.

I'm guessing this might get merged soon and be in the next release(s). What would be the next steps here once the discovery endpoint provides all the supported claims?

Note I don't expect there'd be that many deployments happening to pypi.org from self-hosted, as it's effectively source-available and github.com is where OSS python projects end up hosting their repos. But it would still be best to avoid static credentials where possible.

I see there's some discussion around how to vet self-hosted instances. Our team runs one of the largest self-hosted instances after gitlab.com, with a growing base of ~77k registered users, so would be interested if we can also get on board if there's a pilot project! We're a github [secret scanning partner](https://github.com/github/docs/blob/main/src/secret-scanning/data/public-docs.yml#L3049-L3070) and automatically revoke any Siemens API keys/tokens sent from GitHub, and we could extend that most likely.

Our team's also pretty well versed in GitLab (we contribute regularly, including the initial OIDC provider implementation) so if there's anything missing we could probably add that upstream 👍

**Comment by warsaw** [comment]
> Note I don't expect there'd be that many deployments happening to pypi.org from self-hosted

Actually, at my $work, we want to do exactly that.  We have an internal self-hosted GitLab and want to release packages to PyPI all the time.  We'd like to adopt TP.

**Comment by di** [comment]
> What would be the next steps here once the discovery endpoint provides all the supported claims?

I think we probably want to do what I mentioned above:

> Another option is that we allow-list certain issuers for projects in certain organizations, and manually handle these on a case-by-case basis.

So that would mean creating the views & routes that would allow an organization owner to configure a custom issuer, and updating the Trusted Publishing logic to take this into consideration when publishing.

**Comment by wouterdb** [comment]
Hi All,

The attestation support is great! We have been waiting for this for a long time. 
As the attestations are closely linked to trusted publishing, I'm commenting here.
If this is not the correct place, can you point me to where I should go?

I am very much in favor of allowing self-hosted builds to publish attestations/be trusted publishers because:
1. in the current model, only larger companies can attest. People who choose to self-host (for whatever reason) become untrustworthy in comparison.
2. `setup.py` is an executable, executed both when packaging and when installing, so I can do basically whatever I want in there. The additional security github can provide on providence is very little. github can attest the package came from them, but that is it.
3. Any breach/confusion at the side of the large provider will affect the trust system in a most catastrophic way. Smaller trust sources will have a smaller impact. 

Thanks,
Wouter

**Comment by woodruffw** [comment]
> * in the current model, only larger companies can attest. People who choose to self-host (for whatever reason) become untrustworthy in comparison.

Not exactly: the attestations themselves are done by individual projects. The technique we're using to make attestations verifiable by the index currently relies on Trusted Publishing, meaning that the signing is done by the individual project but the "identity" for the signature is provided by GitHub, GitLab, etc.

In effect, this is a kind of hybrid signing scheme: the project signs for itself, but the signature is intermediated by a trusted identity provider (GitHub, GitLab, etc.) to solve the problem of key/identity discovery in traditional signing schemes. This is explained in the [security considerations](https://docs.pypi.org/attestations/security-model/).

Importantly, providing an attestation is **not evidence of trustworthiness**, and PyPI doesn't describe it as such. All an attestation does is provide a cryptographic link between a PyPI project and its publishing source; that source could be malicious or otherwise untrustworthy. I'll update the docs in a bit to make this even more explicit.

> * `setup.py` is an executable, executed both when packaging and when installing, so I can do basically whatever I want in there. The additional security github can provide on providence is very little. github can attest the package came from them, but that is it.

This goes back to trustworthiness. The provenance is not intended to tell you *a priori* whether `setup.py` is safe to execute or not; you still need to make that decision for yourself; no tool will be able to make it for you with full generality. Provenance not intended to prevent that; it's a part of a larger design, aimed at making it harder for an attacker to *quietly take over* a project or compromise its CI/CD. The goal of provenance is not to tell you whether a project is or isn't trustworthy, but to provide an *automatic, verifiable* way to observe and quantify changes to a project's identity.

> 3\. Any breach/confusion at the side of the large provider will affect the trust system in a most catastrophic way. Smaller trust sources will have a smaller impact.

Yes, this is the basic tradeoff of Trusted Publishing, and it's why PyPI is selective about accepting Trusted Publisher providers that can provide on-call staff and do proper incident response. This is similar to the security calculus for HTTPS: the solution to catastrophic risk is not to allow unauditable CAs into the Web PKI, but to enforce standards on existing CAs *and* force them into transparency schemes, like attestations on PyPI does.

**Comment by tachyons** [comment]
>  GitLab at minimum has an outstanding issue with that configuration being incomplete: https://gitlab.com/gitlab-org/gitlab/-/issues/428061

@woodruffw GitLab Engineer here,   For the context, This mismatch is because we have OIDC discovery URL for multiple OIDC implementations and CI is just one of them. So we can't simply add all claims to discovery URL  till we fix this inconsistency, fixing this this backward compatibility can take some time. 

So I am trying to understand two things

1. How does missing claims block trusted publishing from self managed GitLab instances ?
2. What are the minimum set of claims to added on discovery URL to unblock this effort ?

cc @nejch @nickveenhof

**Comment by di** [comment]
> How does missing claims block trusted publishing from self managed GitLab instances ?

The missing claims in the `.well-known/openid-configuration` list does not block self-hosted GitLab instances if:

a) we can assume the claims we require from the public GitLab instance will always be present in self-hosted instances (and have the same semantics) even if they don't appear in this list
b) we don't need to verify additional claims for self-hosted instances that aren't present in the public instance.

> What are the minimum set of claims to added on discovery URL to unblock this effort ?

I think for now we can just support the same claims for self-hosted instances that we support for the public GitLab instance, which are these:

https://github.com/pypi/warehouse/blob/c66d2d4c1b0bb276484243f701ccfc432450d7cf/warehouse/oidc/models/gitlab.py#L142-L153

These don't _need_ to be in the `.well-known/openid-configuration` list of supported claims for this to work, but it would be good to have some guarantee that these claims are going to be present.

I think if we plan to support verifying any additional claims, especially claims that might only be present on self-hosted instances, we would probably need this configuration to be accurate, especially if they might be different from one self-hosted instance to another (which is what I said [here](https://github.com/pypi/warehouse/issues/15838#issuecomment-2206579784), but I'm still not sure if this would ever be the case).

**Comment by woodruffw** [comment]
@tachyons Sorry for the delay!

I agree with what @di wrote -- the claim discrepancy here isn't a *huge* problem if we can be sure that our current required claims can be enforced on third-party GitLab hosts. The `openid-configuration` would be *nice* for us to have as a ground truth, but some sort of assurance that we can continue to rely on these claims even in third-party setups would IMO suffice 🙂 

More broadly, it would be nice to have the `openid-configuration` show these claims for "shape" reasons -- PyPI (and other TP implementors) need to know what *kind* of provider the host is (e.g. GitLab, GitHub, etc.), and the claim set allows us to infer that. We could allow users to configure this manually (and we'll probably still want them to explicitly declare it for cross-checking purposes), but being able to infer it from the OpenID configuration saves us can help us present better errors/suggestions to users if they enroll their third-party host as the wrong kind by accident.

**Comment by tachyons** [comment]
> a) we can assume the claims we require from the public GitLab instance will always be present in self-hosted instances (and have the same semantics) even if they don't appear in this list

> b) we don't need to verify additional claims for self-hosted instances that aren't present in the public instance.


We have Same code for Gitlab.com and Self managed instances which powers this. So the claims will be the same with two exceptions

1. If a new claim is added in a version x, self managed instances which are old versions of GitLab wouldn't have that claim supported. 

2. If we release a claim which is under a feature flag, it might take one more release to get it on self managed instances. 

Both of these cases are rare and unlikely to block this as the claim change is not frequent and we can't delete existing claims easily as we need to take care of backward compatibility. 


> These don't need to be in the .well-known/openid-configuration list of supported claims for this to work, but it would be good to have some guarantee that these claims are going to be present.

All these claims are shared across self managed and Gitlab.com, also we wouldn't be changing this unless necessary as any change on this will impact several integrations.

**Comment by tachyons** [comment]
We have documented how other platforms including AWS, GCP, and Azure are integrated with [here](https://docs.gitlab.com/ci/cloud_services/)

**Comment by tachyons** [comment]
> More broadly, it would be nice to have the openid-configuration show these claims for "shape" reasons -- PyPI (and other TP implementors) need to know what kind of provider the host is (e.g. GitLab, GitHub, etc.), and the claim set allows us to infer that. We could allow users to configure this manually (and we'll probably still want them to explicitly declare it for cross-checking purposes), but being able to infer it from the OpenID configuration saves us can help us present better errors/suggestions to users if they enroll their third-party host as the wrong kind by accident.

@woodruffw One issue with us that the same discovery URL is shared across multiple OIDC implementations. The claims you see on the discovery URL is the claims support for OIDC as the [login mechanism](https://docs.gitlab.com/integration/openid_connect_provider/). We are planning to extract the discovery URL to separate endpoint, eg

`https://gitlab.example.com/ci/.well-known/openid-configuration`

This is still on the planning phase and not something we can promise on short term. So if we can move ahead without the claims on disovery URL, that would be great. If needed we can add the list of claims @di listed above.

**Comment by nejch** [comment]
Maybe just for future reference, worth adding the documented list of claims provided in the CI ID token by GitLab:

https://docs.gitlab.com/ci/secrets/id_token_authentication/#token-payload

As @tachyons stated these are consistent across all instances and .com except for older instances potentially, but they've been fairly stable.

@tachyons I'm happy to change the upstream MR to advertise only those fields discussed above if the direction is clear 🙂 As I mentioned [here](https://gitlab.com/gitlab-org/gitlab/-/merge_requests/170072#note_2171021320) and [here](https://gitlab.com/gitlab-org/gitlab/-/merge_requests/170072#note_2313415565), the RFC only says that the advertised supported claims _MAY_ be returned in a token, so I don't see an issue with additional claims in the discovery endpoint as it's often cached anyway, but I haven't followed the discussion upstream lately so much. Whatever helps bring this forward happy to change 👍

**Comment by tachyons** [comment]
@nejch I'm waiting for the maintainers response to see if updating the endpoint is essential.

**Comment by woodruffw** [comment]
> updating the endpoint is essential.

From the previous comment: either updating the claims here or giving us some other (standard) way to determine the "shape" of self-hosted GitLab IdPs is a prerequisite for PyPI supporting third-party GitLab instances for trusted publishing. Without that, it's much harder for us to build a clean enrollment flow. 

(I suspect that techniques other than updating the claims will be hard to do while complying with the OIDC standard, so I would definitely prefer those be fixed.)

**Comment by tachyons** [comment]
@woodruffw Many of the claims listed on the discovery URL are not available in the CI. So unfortunately you can't rely on it alone to determine the structure. Only way to get the updated structure of id_token in CI is to use our list [here](https://docs.gitlab.com/ci/secrets/id_token_authentication/#token-payload) as we keep it updated.  

And regarding the implementation wouldn't it be easy if we ask the user if the instance is Gitlab self managed instance or Github self managed instance ?

**Comment by woodruffw** [comment]
> [@woodruffw](https://github.com/woodruffw) Many of the claims listed on the discovery URL are not available in the CI. So unfortunately you can't rely on it alone to determine the structure. Only way to get the updated structure of id_token in CI is to use our list [here](https://docs.gitlab.com/ci/secrets/id_token_authentication/#token-payload) as we keep it updated.

It's completely fine for integration purposes if the list of claims is a *superset* of the ones available on the CI's OIDC IdP -- to my understanding, that's how the OIDC discovery endpoint is meant to work in the first place (it's not a *guarantee* that the claims are present).

From the OIDC discovery spec:

> Metadata Name: claims_supported
Metadata Description: JSON array containing a list of the Claim Names of the Claims that the OpenID Provider MAY be able to supply values for
Change Controller: OpenID Foundation Artifact Binding Working Group - openid-specs-ab@lists.openid.net
Specification Document(s): [Section 3](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata) of this specification

Ref: https://openid.net/specs/openid-connect-discovery-1_0.html#MetadataContents

In other words, there's no compatibility risk if you add the CI claims to the current list. There's no need to remove the other claims or split them in any way, PyPI can simply check for the subset it needs.


> And regarding the implementation wouldn't it be easy if we ask the user if the instance is Gitlab self managed instance or Github self managed instance ?

This results in a pretty confusing/suboptimal user and administrator experience if the user gives us the wrong "color" of IdP at enrollment time: we'd silently accept it and only fail at *publish* time, which could be indefinitely later.

**Comment by nejch** [comment]
That's also my understanding of the spec, hence the naming of the `claims_supported` field (as opposed to something like `claims_provided`), but I have no strong opinion on this.

**Comment by miketheman** [comment]
> Only way to get the updated structure of id_token in CI is to use our list [here](https://docs.gitlab.com/ci/secrets/id_token_authentication/#token-payload) as we keep it updated.

Thanks for the link! As I read through, I noticed entries that may state that they are provided "Always" - which is good to know.
Some have no "when added" and others contain "Introduced in GitLab XX.Y" type notes, e.g. `job_project_id`

When interacting with GitLab.com, we have some degree of certainty of the version running, but since nothing in the `.well-known/openid-configuration` nor the generated `id_tokens` contain GitLab version information, how would you recommend we rely on the "always"-ness of these keys?
I recognize that the version information is contained in an authenticated metadata API call, but is there any other place I can reliably detect this version data?

**Comment by di** [comment]
Given the change in #18814 which was merged, I wanted to note that we'll also need a better way to map a self-hosted issuer URL to a specific publisher type, since we currently follow [this](https://github.com/pypi/warehouse/blob/ee2fe4154807fdf21823213939db33e60349ca8e/warehouse/oidc/utils.py#L69-L96) process:

1. the issuer URL maps to a publisher class
2. the class is used to make sure all the necessary claims exist
3. the class is used to look up a specific publisher based on the claims

For GitLab publishers that have an issuer other than `https://gitlab.com`, the first step there breaks down, and I don't think we'll want to have to update this mapping in code to support all self-hosted instances.

**Comment by tachyons** [comment]
https://gitlab.com/.well-known/openid-configuration Now has the additional scopes listed above.

**Comment by miketheman** [comment]
@tachyons Thanks for the update, that's great news!
I'm not sure how y'all's feature flags work, but is this rolled out to 100% of all requests?
From what I can gather, you're scheduling this to be part of the 18.5 release, correct?

If we can begin relying on the data from the `claims_supported` key, we can begin developing stronger validations, and for self-managed instance provisioning we'd require users to have 18.5 as a minimum version?

**Comment by tachyons** [comment]
@miketheman It is enabled for Gitlab.com. We missed the cut off to enable this by default for 18.5. So the feature will be released under a feature flag in 18.5 and enabled by default in 18.6

**Comment by miketheman** [comment]
TODO: Clarify what the "namespace" ought to be - sometimes it's `foo` and sometimes it's `foo/bar`, resulting in URLs that may not work when appended with the project name, a la `foo/bar/bar`

**Comment by matthewfeickert** [comment]
There's questions over on https://github.com/scientific-python/upload-nightly-action/issues/158 about if it would be possible to use Trusted publishers with anaconda.org, but I think the answer is a similar number of things would need to get addressed for anaconda.org to become a [verified platform](https://docs.pypi.org/project_metadata/#via-trusted-publishing). Is that correct?

**Comment by di** [comment]
Thanks for the pointer! The docs at https://docs.pypi.org/trusted-publishers/internals/#how-do-i-become-a-trusted-publishing-provider have all the details, and I've left a comment on that issue to that effect.

**Comment by maxnoe** [comment]
Hi,

We have been testing trusted publishing from our GitLab instance to our PyPI org now and while it works great for existing projects, it seems the "Pending Publisher" form is missing the dropdown for the GitLab instance.

Existing project:

<img width="764" height="1026" alt="Image" src="https://github.com/user-attachments/assets/4cc6bc60-3653-4986-818a-688c1fb1a82d" />

Pending Publisher:

<img width="764" height="1115" alt="Image" src="https://github.com/user-attachments/assets/6978b9ab-9443-4bd9-acdd-41c342818a8a" />

**Comment by miketheman** [comment]
@maxnoe That's odd! Can you confirm that you are looking at the Organization-level Publishing page, vs the account-level Publishing page? The path should look like `/manage/organization/<org_name>/publishing/`
instead of account-level which is `/manage/account/publishing/`

You may have to click on "Your Organizations" -> Select one of the organizations and click "Manage" -> and then a sidebar tab for "Publishing" for the Organization scope should be there. Let me know if that can't be found.

**Comment by maxnoe** [comment]
Sorry, my bad, indeed I was in my account and not the org.

I was confused by this in the docs here and then didn't check again properly

<img width="982" height="314" alt="Image" src="https://github.com/user-attachments/assets/88a5b826-b467-44a3-a56d-7d44534e5b48" />

:

**Comment by maxnoe** [comment]
I understand now better what happened.

I clicked "My Organizations" and then "Publishing". But that brings me to my account's publishing page.
The correct flow is "My Organizations" → The organizaiton's "Manage" button in the list → "Publishing"

**Comment by miketheman** [comment]
The public docs haven't been updated yet, pending the beta

**Comment by maxnoe** [comment]
Hi,

Today we get errors like these in our publishing pipelines, in projects where the trusted publishing previously worked and that still have trusted publishing defined on pypi:

```
$ twine upload ${PYTHON_PROJECT_DIR}/dist/*
Uploading distributions to https://upload.pypi.org/legacy/
WARNING  Got OIDC token for audience pypi                                       
ERROR    TrustedPublishingFailure: The token request failed; the index server   
         gave the following reasons:                                            
                                                                                
         * `invalid-publisher`: valid token, but no corresponding publisher (JWT
         issuer 'https://gitlab.cta-observatory.org/' does not match expected    
         issuer 'https://gitlab.com/')
```

The PyPI settings for these projects show e.g.:

<img width="546" height="132" alt="Image" src="https://github.com/user-attachments/assets/c6ee8d88-691a-4909-b66f-6b52bafc11dc" />

which is the correct config, and which as worked before:

<img width="450" height="185" alt="Image" src="https://github.com/user-attachments/assets/4cacc941-3ac7-41c8-ae6f-df58932ca705" />

<img width="807" height="193" alt="Image" src="https://github.com/user-attachments/assets/47f58adf-5a0e-404b-92d0-b00d50071af0" />

**Comment by maxnoe** [comment]
Pretty sure this commit broke it: https://github.com/pypi/warehouse/commit/a44788571c69e4f1d41c53b076673e9fd5a07061

**Comment by miketheman** [comment]
Hey @maxnoe - sorry about that. I believe https://github.com/pypi/warehouse/pull/19718 ought to correct the behavior in a more sustainable manner.
TL,DR: the fix applied previously was to project against a PyPI Admin from misconfiguring a service that wouldn't be evaluated correctly. This check is should be more robust.

**Comment by maxnoe** [comment]
We can publish again, thanks a lot.

**Comment by miketheman** [comment]
Now that publishing works as intended, I'm going to close this one. 🥳 

Any new issues should use a New Issue.
